File: Protocol/_theoreme_sim/src/propagation/decayModel.js
/* Purpose: Apply optional network-wide decay for charges and neighbor estimate tables. */

/**
 * @param {Map<number,{qTotal:number}>} nodes
 * @param {Map<number,Map<number,number>>} estimates
 * @param {number} factor
 */
export function applyGlobalDecay(nodes, estimates, factor) {
  const clamped = Math.min(Math.max(factor, 0.01), 1);

  for (const node of nodes.values()) {
    node.qTotal *= clamped;
  }

  for (const neighborMap of estimates.values()) {
    for (const [neighborId, value] of neighborMap.entries()) {
      neighborMap.set(neighborId, value * clamped);
    }
  }
}

File: Protocol/_theoreme_sim_py/src/propagation/decay_model.py
"""Purpose: Apply optional global decay to charges and neighbor estimate tables."""

from __future__ import annotations

from src.utils.math_utils import clamp


def apply_global_decay(nodes: dict, estimates: dict[int, dict[int, float]], factor: float) -> None:
    clamped = clamp(float(factor), 0.01, 1)

    for node in nodes.values():
        node.q_total *= clamped

    for neighbor_map in estimates.values():
        for neighbor_id, value in list(neighbor_map.items()):
            neighbor_map[neighbor_id] = float(value) * clamped
